### PR TITLE
Add integration test for GKE

### LINKED ISSUE #96: Test the integration with Google Container Engine
We should add an integration test for the deployment to Google Container Engine (based on [Integration Frameworks](https://github.com/GoogleCloudPlatform/runtimes-common/tree/master/integration_tests))

### PR DESCRIPTION
**Pull request by cassand**
Add a script to setup environment for integration test on GKE using [runtimes-common integration testing framework](https://github.com/GoogleCloudPlatform/runtimes-common/tree/master/integration_tests)

The script does the following steps: 
- Build a test application into a docker image
- Upload the image to Google Container Registry
- Create a Kubernetes cluster on GKE
- Deploy the application
- Run the integration test

It assumes that kubectl executable is available.

This PR fixes #96

### CODE REVIEW

File: scripts/gke_integration_test.sh [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+popd
+
+# deploy to app engine
+pushd ${deployDir}
+export STAGING_IMAGE=${imageUnderTest}
+envsubst < "Dockerfile.in" > "Dockerfile"
+export TESTED_IMAGE=${imageUrl}
+envsubst < "openjdk-spring-boot.yaml.in" > "openjdk-spring-boot.yaml"
+
+echo "Deploying image to Google Container Engine..."
```
Comment (meltsufin) [inline scripts/gke_integration_test.sh]: "Deploying image to Google Container **Registry**..."

File: scripts/gke_integration_test.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+docker build -t "$imageName" .
+docker tag "$imageName" "$imageUrl"
+gcloud docker -- push gcr.io/${projectName}/openjdk-gke-integration
```
Comment (meltsufin) [inline scripts/gke_integration_test.sh]: @aslo Should we do all these steps or use the builder instead? (In any case, this could be a separate PR.)

Comment (aslo) [inline scripts/gke_integration_test.sh]: It's worth thinking about. One reason why it might not be a good idea - these tests require using a custom base image, which would require overriding the default builder to point to the base image under test. 

I think if we want to test openjdk runtime candidates with the builder pipeline, that would belong in the builder's integration tests.

File: scripts/integration_test.sh [hunk 3]
```diff
@@ -14,40 +14,17 @@
...
-DEPLOYED_APP_URL="http://$(gcloud app describe | grep defaultHostname | awk '{print $2}')"
-echo "Running integration tests on application that is deployed at $DEPLOYED_APP_URL"
-
-# run in cloud container builder
-gcloud container builds submit \
-  --config $dir/integration_test.yaml \
-  --substitutions "_DEPLOYED_APP_URL=$DEPLOYED_APP_URL" \
-  $dir
+${dir}/ae_integration_test.sh ${imageUnderTest}
+${dir}/gke_integration_test.sh ${imageUnderTest}
```
Comment (meltsufin) [inline scripts/integration_test.sh]: Maybe add a final line here saying:
App Engine integration tests: PASSED/FAILED
Container Engine integration tests: PASSED/FAILED

File: DEVELOPING.md [hunk 4]
```diff
@@ -38,6 +38,16 @@ to a gcr.io repository before the integration tests can run.
 ```bash
 $ RUNTIME_IMAGE=gcr.io/my-project-id/openjdk:my-tag
 $ gcloud docker -- push $RUNTIME_IMAGE
-$ ./scripts/integration_test.sh $RUNTIME_IMAGE
```
Comment (meltsufin) [inline DEVELOPING.md]: Let's leave this in for the case when people want to run both integration tests.
Also, in the the paragraph above, mention GKE in addition to App Engine.

File: DEVELOPING.md [hunk 5]
```diff
@@ -38,6 +38,16 @@ to a gcr.io repository before the integration tests can run.
...
 ```
 
+You have the possibility to either run the tests on App Engine or on Google Container Engine.
```
Comment (meltsufin) [inline DEVELOPING.md]: Then here just say that you have the option to run the AE and GKE tests individually like this:

File: scripts/integration_test.sh [hunk 6]
```diff
@@ -17,14 +17,29 @@
 # Setup environment for CI.
 set -e
 
-dir=`dirname $0`
+readonly dir=`dirname $0`
+
+GetStatusOfContainerBuild () {
```
Comment (meltsufin) [inline scripts/integration_test.sh]: Can we just rely on the exit code of the script instead? I believe it's 0 when pass, and non-zero when failed.

File: scripts/integration_test.sh [hunk 7]
```diff
@@ -17,14 +17,29 @@
...
-${dir}/ae_integration_test.sh ${imageUnderTest}
-${dir}/gke_integration_test.sh ${imageUnderTest}
+AE_OUTPUT=$(${dir}/ae_integration_test.sh ${imageUnderTest})
+GKE_OUTPUT=$(${dir}/gke_integration_test.sh ${imageUnderTest})
+
+AE_STATUS=$(GetStatusOfContainerBuild "$AE_OUTPUT")
+GKE_STATUS=$(GetStatusOfContainerBuild "$GKE_OUTPUT")
+
+echo "App Engine integration tests: $AE_STATUS"
+echo "Container Engine integration tests: $GKE_STATUS"
```
Comment (meltsufin) [inline scripts/integration_test.sh]: Also, I think we want to make sure the script return non-zero exit code, if either of the tests failed.

File: scripts/integration_test.sh [hunk 8]
```diff
@@ -14,40 +14,33 @@
...
+${dir}/ae_integration_test.sh ${imageUnderTest}
+AE_OUTPUT=$?
+AE_STATUS=$(GetStatusOfContainerBuild "$AE_OUTPUT")
+
+${dir}/gke_integration_test.sh ${imageUnderTest}
+GKE_OUTPUT=$?
+GKE_STATUS=$(GetStatusOfContainerBuild "$GKE_OUTPUT")
+
+echo "App Engine integration tests: $AE_STATUS"
+echo "Container Engine integration tests: $GKE_STATUS"
```
Comment (meltsufin) [inline scripts/integration_test.sh]: I think we need an `exit` instruction here that returns `0` if both passed or `1` otherwise.

Comment (aslo) [inline scripts/integration_test.sh]: I think that's not necessary, since the script has `set -e` at the top. It will exit immediately if any command exits with non-zero status. 

If that's not what @cassand intends here, then `set -e` should be removed and we can compute the exit code of this script based on how `ae_integration_test` and `gke_integration_test` finish.

Comment (meltsufin) [inline scripts/integration_test.sh]: I totally missed the `set -e` part. Thanks for pointing it out! I think we just want to disable `set -e` right before calling out to the test scripts by using `set +e`.

Comment (cassand) [inline scripts/integration_test.sh]: It seems fine that way, however, I am not sure that the command `gcloud container builds` exit with 1 if the remote build fail, in which case it may be useful to exit based on the status.

Comment (cassand) [inline scripts/integration_test.sh]: I confirm that the command `gcloud container builds` exit with a non zero exit code in case of failure

File: scripts/integration_test.sh [hunk 9]
```diff
@@ -14,40 +14,17 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# Setup environment for CI.
```
Comment (meltsufin) [inline scripts/integration_test.sh]: "Exit as soon as any of the commands in the script fail"

File: scripts/integration_test.sh [hunk 10]
```diff
@@ -15,6 +15,8 @@
 # limitations under the License.
 
 # Setup environment for CI.
+# This script exit as soon as any of the commands in the integration scripts fail
```
Comment (meltsufin) [inline scripts/integration_test.sh]: exit -> exits
